rational.go

```diff
@@ -53,3 +64,46 @@ func (r Rational) String() string {
 func (r Rational) Invert() Rational {
 	return NewRational(r.Den(), r.Num())
 }
+
+type rational struct {
+	Num int
+	Den int
+}
+
+func (r *Rational) UnmarshalJSON(d []byte) error {
+	var rs rational
+	if err := json.Unmarshal(d, &rs); err != nil {
+		return err
+	}
+	r.SetNum(rs.Num)
+	r.SetDen(rs.Den)
+	return nil
+}
+
+func (r Rational) MarshalJSON() ([]byte, error) {
+	return json.Marshal(rational{Num: r.Num(), Den: r.Den()})
+}
+
+func (r Rational) MarshalText() ([]byte, error) { return ([]byte)(r.String()), nil }
+func (r *Rational) UnmarshalText(d []byte) error {
+	s := string(d)
+	if s == "" {
+		r.SetNum(0)
```

Comment:
Can you do `*r = NewRational(0, 0)` instead?

Also you need to return `nil` here.